### ISSUE TITLE
Memory leak in DynamoDB Query

### ISSUE THREAD

**Issue by marcio-oliveira-cx** [aws/aws-sdk-net#2479]
### Describe the bug

We are facing a memory leak requesting a lot of data by QueryAsync which leads to ECS Task to fail by OutOfMemory exception.
I have a POC that show this, which I request 1000 queries with 1 second interval which leads to not clear memory after all requests being done.

### Expected Behavior

RAM should be released after all queries being retrieved.

### Current Behavior

I have 1k requests that will do a QueryAsync to a DynamoDB table and will retrieve 150mb of data but RAM will increase to near 2GB of RAM and does not clear after that most of the times (sometimes does).
This requests are being parallelized to have a better performance.
If I stop this requests the memory is not cleared.

The next image shows the memory RAM of my POC after a few seconds.
![Memory leak](https://user-images.githubusercontent.com/97117269/200790107-e2f629a7-7d30-468b-9b04-2f4033ff6a89.png)

### Reproduction Steps

Send multiple requests with large amount of response data in intervals of 1 second.

### Possible Solution

There is no possible solution.

### Additional Information/Context

Our application is hosted in ECS container with 8GB of RAM and when a lot of requests are being sent to our application it will increase memory until it throws an OutOfMemory exception and Stops.

### AWS .NET SDK and/or Package version used

AWSSDK.DynamoDBv2 3.7.100.8

### Targeted .NET Platform

.NET 6

### Operating System and version

Windows 11, Linux

**Comment by marcio-oliveira-cx** [comment]
Hi @ashishdhingra 
Yes, I can provide you a sample code.
I'm doing a Background Task with 1 second of interval that does 1k requests to Dynamo which I split into 400 requests in Parallel.
The requests are made by a Query to the GSI.

protected override async Task ExecuteAsync(CancellationToken stoppingToken)
{
    var timer = new PeriodicTimer(TimeSpan.FromSeconds(1));

    do
    {
        await hashKeys.ParallelForEachAsync(400, async hashKey =>
        {
            await dynamoDbContext.QueryAsync<FileHashDao>(hashKey, operationConfig).GetRemainingAsync(stoppingToken);
        });

    } while (await timer.WaitForNextTickAsync(stoppingToken) && !stoppingToken.IsCancellationRequested);
}

public static Task ParallelForEachAsync<T>(this IEnumerable<T> source, int degreeOfParallelism, Func<T, Task> action)
{
    async Task AwaitPartition(IEnumerator<T> partition)
    {
        using (partition)
        {
            while (partition.MoveNext())
            { await action(partition.Current); }
        }
    }
    return Task.WhenAll(
        Partitioner
            .Create(source)
            .GetPartitions(degreeOfParallelism)
            .AsParallel()
            .Select(AwaitPartition));
}

**Comment by p-brito** [comment]
@ashishdhingra I also noticed the same thing, memory is quickly drained when queries are against the secondary table.

**Comment by p-brito** [comment]
Here is the stack trace, if it helps.
`System.OutOfMemoryException: Exception of type 'System.OutOfMemoryException' was thrown.
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetException(Exception exception)
   at Amazon.Runtime.RetryPolicy.RetryAsync(IExecutionContext executionContext, Exception exception)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Amazon.Runtime.RetryPolicy.RetryAsync(IExecutionContext executionContext, Exception exception)
   at Amazon.Runtime.Internal.RetryHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.CallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.CallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.ErrorCallbackHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.Runtime.Internal.MetricsHandler.InvokeAsync[T](IExecutionContext executionContext)
   at Amazon.DynamoDBv2.DocumentModel.Search.GetNextSetHelperAsync(CancellationToken cancellationToken)
   at Amazon.DynamoDBv2.DocumentModel.Search.GetRemainingHelperAsync(CancellationToken cancellationToken)
   at Amazon.DynamoDBv2.DataModel.AsyncSearch`1.GetRemainingAsync(CancellationToken cancellationToken)
   at Lumo.DynamoDb.Client.DynamoDbClient.Query[T](String hashKeyFieldName, Primitive hashKeyFieldValue, QueryOptions queryOptions)`

**Comment by dscpinheiro** [comment]
Not sure if if it's possible, but would you be able to share more details about the data model?

I've been trying to run the provided code, but I'm using the sample from the Dynamo documentation (https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/SampleData.html) and memory doesn't spike as much. 

I tried adding more attributes to each item (so they're close to the 400kb limit), but it's still a relatively simple / small schema. Here's my setup:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="AWSSDK.DynamoDBv2" Version="3.7.100.19" />
  </ItemGroup>
</Project>
```

```csharp
using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.DataModel;
using System.Collections.Concurrent;

internal class Program
{
    static readonly AmazonDynamoDBClient _ddbClient = new();
    static readonly DynamoDBContext _ddbContext = new(_ddbClient);

    static async Task Main()
    {
        using var cts = new CancellationTokenSource();
        await ExecuteAsync(cts.Token);
    }

    static async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        var timer = new PeriodicTimer(TimeSpan.FromSeconds(1));
        var hashKeys = new List<string> { "User A", "User B", "Non-existent User" };
        var operationConfig = new DynamoDBOperationConfig { IndexName = "PostedBy-Message-index" };

        do
        {
            await hashKeys.ParallelForEachAsync(400, async hashKey =>
            {
                var response = await _ddbContext.QueryAsync<Reply>(hashKey, operationConfig).GetRemainingAsync(stoppingToken);
                Console.WriteLine($"Found {response.Count} items for PK: {hashKey}");
            });
        } while (await timer.WaitForNextTickAsync(stoppingToken) && !stoppingToken.IsCancellationRequested);
    }
}

static class Helpers
{
    public static Task ParallelForEachAsync<T>(this IEnumerable<T> source, int degreeOfParallelism, Func<T, Task> action)
    {
        async Task AwaitPartition(IEnumerator<T> partition)
        {
            using (partition)
            {
                while (partition.MoveNext())
                { 
                    await action(partition.Current); 
                }
            }
        }

        return Task.WhenAll(
            Partitioner
                .Create(source)
                .GetPartitions(degreeOfParallelism)
                .AsParallel()
                .Select(AwaitPartition)
        );
    }
}

[DynamoDBTable("Reply")]
public class Reply
{
    [DynamoDBHashKey]
    public string Id { get; set; }

    [DynamoDBRangeKey]
    public DateTime ReplyDateTime { get; set; }

    public string Message { get; set; }

    [DynamoDBProperty("LastPostedBy")]
    public string PostedBy { get; set; }
    
    [DynamoDBVersion]
    public int? Version { get; set; }
}
```

We do have a few other performance related issues about DynamoDB (https://github.com/aws/aws-sdk-net/labels/module%2Fdynamodb), unfortunately most of them would require a refactor of the module.

**Comment by marcio-oliveira-cx** [comment]
@dscpinheiro The schema seems fine.
My schema are just like this in the main table but on the GSI it flips the HashKey and RangeKey, so I just do queries by the HashKey of the GSI:

`
[DynamoDBTable("TestEntities")]
public class TestEntity
{
    [DynamoDBHashKey]
    public string Id { get; set; }

    [DynamoDBRangeKey]
    public string Key { get; set; }
    
    public string Description { get; set; }
    
    public DateTime ReleaseDate { get; set; }

    public DateTime InsertionTime { get; init; } = DateTime.Now;
}`

Are you retrieving a lot of data?
In my tests if I deserialize all the objects that I get to a text file I get like 150mb of data from DynamoDB, so I think it is related to how much data it is being retrieved from the database.

**Comment by FrankRua** [comment]
> @dscpinheiro The schema seems fine. My schema are just like this in the main table but on the GSI it flips the HashKey and RangeKey, so I just do queries by the HashKey of the GSI:
> 
> ```
> [DynamoDBTable("TestEntities")]
> public class TestEntity
> {
>     [DynamoDBHashKey]
>     public string Id { get; set; }
> 
>     [DynamoDBRangeKey]
>     public string Key { get; set; }
>     
>     public string Description { get; set; }
>     
>     public DateTime ReleaseDate { get; set; }
> 
>     public DateTime InsertionTime { get; init; } = DateTime.Now;
> }
> ```
> 
> Are you retrieving a lot of data? In my tests if I deserialize all the objects that I get to a text file I get like 150mb of data from DynamoDB, so I think it is related to how much data it is being retrieved from the database.

I am trying to do something similar, have you found a work around?

**Comment by marcio-oliveira-cx** [comment]
> > @dscpinheiro The schema seems fine. My schema are just like this in the main table but on the GSI it flips the HashKey and RangeKey, so I just do queries by the HashKey of the GSI:
> > ```
> > [DynamoDBTable("TestEntities")]
> > public class TestEntity
> > {
> >     [DynamoDBHashKey]
> >     public string Id { get; set; }
> > 
> >     [DynamoDBRangeKey]
> >     public string Key { get; set; }
> >     
> >     public string Description { get; set; }
> >     
> >     public DateTime ReleaseDate { get; set; }
> > 
> >     public DateTime InsertionTime { get; init; } = DateTime.Now;
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Are you retrieving a lot of data? In my tests if I deserialize all the objects that I get to a text file I get like 150mb of data from DynamoDB, so I think it is related to how much data it is being retrieved from the database.
> 
> I am trying to do something similar, have you found a work around?

Hey @FrankRua, did not found a work around.
What I have tried is to try to cap the number of items retrieved but that's not a final solution.

**Comment by FrankRua** [comment]
Is there any update on this issue? This seems crazy nobody has addressed it in months!

**Comment by normj** [comment]
Can the people on the thread checkout this blog post to see if it helps alleviate the memory issues you are seeing. https://aws.amazon.com/blogs/developer/configuring-net-garbage-collection-for-amazon-ecs-and-aws-lambda/

**Comment by marcio-oliveira-cx** [comment]
Hi @normj , I tested what was on that thread and from the services that I tested, it stopped having Out Of Memory issues in the tasks.
With a hard memory limit set on the container looks like that the container will not have spikes near 100% used memory, so in this way, Garbage Collector will work as expected.
My question is why this is an optional setting? 
Because without this maybe all services using some demanding memory issues will need to have it.

But the issue persists because I can see that the memory will not decrease in the ECS.

**Comment by normj** [comment]
@marcio-oliveira-cx My understanding is many languages do understand the hierarchy of `cgroups` that are used in ECS. This was limitation in .NET itself. We are working with Microsoft to get it to understand the hierarchy as well. When that is done then setting this value for .NET would be redundant.

**Comment by normj** [comment]
I'm closing this issue because I feel pretty strongly that the blog post [mentioned above](https://github.com/aws/aws-sdk-net/issues/2479#issuecomment-1478244218) addresses the issue. We have not been able to reproduce the problem outside of environments that didn't have the workarounds in the blog post applied. If anybody has a different issue that is not addressed by the blog post please open a new issue. 

We are working with Microsoft so that hopefully with .NET 8 the workarounds mentioned in the blog post will be automatic.
